base-notebook/start.sh

```diff
@@ -11,140 +12,185 @@ else
+    # Optionally ensure the desired user get filesystem ownership of it's home
+    # folder and/or additional folders
+    if [[ "${CHOWN_HOME}" == "1" || "${CHOWN_HOME}" == "yes" ]]; then
+        echo "Ensuring /home/${NB_USER} is owned by ${NB_UID}:${NB_GID} ${CHOWN_HOME_OPTS:+chown options: ${CHOWN_HOME_OPTS}}"
         # shellcheck disable=SC2086
         chown ${CHOWN_HOME_OPTS} "${NB_UID}:${NB_GID}" "/home/${NB_USER}"
     fi
     if [ -n "${CHOWN_EXTRA}" ]; then
         for extra_dir in $(echo "${CHOWN_EXTRA}" | tr ',' ' '); do
-            echo "Changing ownership of ${extra_dir} to ${NB_UID}:${NB_GID} with options '${CHOWN_EXTRA_OPTS}'"
+            echo "Ensuring ${extra_dir} is owned by ${NB_UID}:${NB_GID} ${CHOWN_HOME_OPTS:+(chown options: ${CHOWN_HOME_OPTS})}"
             # shellcheck disable=SC2086
             chown ${CHOWN_EXTRA_OPTS} "${NB_UID}:${NB_GID}" "${extra_dir}"
         done
     fi
 
-    # Change UID:GID of NB_USER to NB_UID:NB_GID if it does not match
-    if [ "${NB_UID}" != "$(id -u "${NB_USER}")" ] || [ "${NB_GID}" != "$(id -g "${NB_USER}")" ]; then
-        echo "Set user ${NB_USER} UID:GID to: ${NB_UID}:${NB_GID}"
-        if [ "${NB_GID}" != "$(id -g "${NB_USER}")" ]; then
-            groupadd -f -g "${NB_GID}" -o "${NB_GROUP:-${NB_USER}}"
-        fi
-        userdel "${NB_USER}"
-        useradd --home "/home/${NB_USER}" -u "${NB_UID}" -g "${NB_GID}" -G 100 -l "${NB_USER}"
-    fi
-
-    # Enable sudo if requested
-    if [[ "${GRANT_SUDO}" == "1" || "${GRANT_SUDO}" == 'yes' ]]; then
-        echo "Granting ${NB_USER} sudo access and appending ${CONDA_DIR}/bin to sudo PATH"
-        echo "${NB_USER} ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/notebook
-    fi
+    # Update potentially outdated environment variables since image build
+    export XDG_CACHE_HOME=/home/$NB_USER/.cache
 
     # Add ${CONDA_DIR}/bin to sudo secure_path
     sed -r "s#Defaults\s+secure_path\s*=\s*\"?([^\"]+)\"?#Defaults secure_path=\"\1:${CONDA_DIR}/bin\"#" /etc/sudoers | grep secure_path > /etc/sudoers.d/path
 
-    # Exec the command as NB_USER with the PATH and the rest of
-    # the environment preserved
+    # Optionally grant passwordless sudo rights for the desired user
+    if [[ "$GRANT_SUDO" == "1" || "$GRANT_SUDO" == 'yes' ]]; then
+        echo "Granting $NB_USER passwordless sudo rights!"
+        echo "$NB_USER ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers.d/added-by-start-script
+    fi
+
+    # NOTE: This hook is run as the root user!
     run-hooks /usr/local/bin/before-notebook.d
-    echo "Executing the command:" "${cmd[@]}"
-    exec sudo -E -H -u "${NB_USER}" PATH="${PATH}" XDG_CACHE_HOME="/home/${NB_USER}/.cache" PYTHONPATH="${PYTHONPATH:-}" "${cmd[@]}"
+
+    echo "Running as ${NB_USER}:" "${cmd[@]}"
+    exec sudo --preserve-env --set-home --user "${NB_USER}" \
+        PATH="${PATH}" XDG_CACHE_HOME="/home/${NB_USER}/.cache" \
```

Comment:
You have already set XDG_CACHE_HOME, do you really need it here as well?
Anyway, you could pass it as XDG_CACHE_HOME="${XDG_CACHE_HOME}"

Reply:
Looking at [past notes](https://github.com/jupyter/docker-stacks/pull/1052/files#diff-41f90d7afcdae13f8516195e078d0a203972b5c5105851eaecfc0a98e9739107R147-R172), I can verify that we don't need to specify XDG_CACHE_HOME explicitly here.

`sudo sudo -V` inspection makes me conclude PATH and PYTHONPATH will need to be set here though.

Reply:
@consideRatio, sorry for [reverting your past notes](https://github.com/jupyter/docker-stacks/commit/4ef8a7816dbb3f8b6f3c69c8c2a640db001a45aa), :pray:  it's just a temporary measure to push through this PR.

I'm wondering what is the difference between declaring it here and exporting it [here](https://github.com/jupyter/docker-stacks/blob/1ce72b55e273f1c566382125a6a0677534fcd502/base-notebook/start.sh#L131). Exporting has the advantage that it's available for the `before-notebook.d` hooks.

Reply:
```suggestion
        PATH="${PATH}" \
```

I forgot to submit my review notes, it should be removed from here, and we should have it updated quickly before running hooks etc.

Reply:
@maresb no worries at all! I'm very appreciative of your ability to make this PR not try to do too much because it's quite tricky to review this as it is with moved parts etc - due to me ^^

Reply:
Should we remove `XDG_CACHE_HOME` as in your suggested change in this PR or a separate one?

BTW, what's actually using `XDG_CACHE_HOME`? I noticed that `/home/jovyan` has `.cache/` but I wasn't sure where it comes from. Is it used by one of the installers? I'm surprised it has to be set in the first place since it should default to `~/.cache`. (Maybe we should add a comment to explain.)

Reply:
I think it's something about matplotlib...

+1 for removing it here, it isn't changing any behavior, just refactoring it to be shipped as part of the `--preserve-env` flag as it is now set already with an export statement above.